rfcs/20190815-tfdbg-v2-callbacks.md

```diff
@@ -0,0 +1,356 @@
+    eager execution of an op. The callback should provide visibility into the
+    input and output `EagerTensor`s of the execution. If so desired, the
+    callback may _override_ the output `EagerTensor`s and thereby transparently
+    affect downstream eager execution.
+  - **Capability B**: The ability to intercept the creation of symbolic ops
+    during function-to-graph conversion in `tf.function`, including the cases
+    assisted by [AutoGraph](https://www.tensorflow.org/guide/autograph). This
+    will form the basis for simulated stepping through lines of
+    `tf.function`-decorated Python functions to assist debugging of graph
+    construction in TF2. Like in Capability A, the callbacks should be able to
+    override the output symbolic tensors of the op in order to affect all
+    downstream graph ops.
+  - **Capability C**: Similar to Capability A above, we want the ability
+    to intercept the runtime execution of FuncGraphs. Note that this
+    requirement could be folded into Capability A, if FuncGraphs are
+    regarded as a special type of op.
+
+Capability B will enable the interception of TF ops executing inside FuncGraphs
+at _runtime_. Although this is not listed explicitly as a desired capability, it
+is critical to runtime debugging in tfdbg v2. Similar to Capability A above, we
+want access to all the intermediate tensor values computed in a FuncGraph.
+
+### Design: Debug Callbacks for Op Instrumentation
+
+The following API meets the requirements listed above. The code below shows the
+API and the detailed signature and semantics of the callbacks that can be passed
+to the API.
+
+```python
+# Exported publicly as: tf.debugging.op_callback()
+def op_callback(callback_fn):
+  """Intercepts op execution and op creation.
+
+  The `callback_fn` will be invoked immediately after any of the three types
+  of events:
+    - The execution of an op under eager mode,
+    - The execution of a FuncGraph under eager mode,
+    - The creation of an op during graph construction (e.g., in
+      @tf.function-decorated Python functions.)
+
+  Args:
+    A callback_fn that has the following signature:
+      def callback_fn(op_type,
+                      inputs,
+                      attrs,
+                      outputs,
+                      op_name=None,
+                      graph=None):
+        # op_type: The type of the op. E.g., "MatMul".
+        #          For the special case of FuncGraph execution, op_type
+        #          takes the special value `tf.Graph`.
+        # inputs: (`tuple` of `Tensor`s) Input tensors to the op or the
+        #         FuncGraph.
+        #         In eager execution, these are `EagerTensor`s.
+        #         In graph construction, these are non-eager `Tensor`s
+        #         that form the inputs to the just-created op.
+        # attrs: The attributes of the op or FuncGraph of which the execution
+        #        or creation caused the current invocation of the callback.
```

Comment:
I don't understand this-- can you explain further?

Reply:
These are the attributes of the op. It is a tuple of alternating attribute key and attribute value, e.g., 

```
('adjoint_a', False, 'adjoint_b', False, 'Tindices', 9L, 'T', 1L, 'T', 1L)
```

I beefed up the doc string here to clarify it.